pkl-cli/src/main/kotlin/org/pkl/cli/CliFormatterCommand.kt

```diff
@@ -15,48 +15,162 @@
+import kotlin.math.max
 import org.pkl.commons.cli.CliBaseOptions
 import org.pkl.commons.cli.CliCommand
+import org.pkl.commons.cli.CliTestException
 import org.pkl.formatter.Formatter
 import org.pkl.formatter.GrammarVersion
 import org.pkl.parser.GenericParserError
 
-abstract class CliFormatterCommand
+class CliFormatterCommand
 @JvmOverloads
 constructor(
-  options: CliBaseOptions,
-  protected val paths: List<Path>,
-  protected val grammarVersion: GrammarVersion,
-  protected val consoleWriter: Writer = System.out.writer(),
-) : CliCommand(options) {
-  protected fun format(file: Path, contents: String): Pair<String, Int> {
+  private val paths: List<Path>,
+  private val grammarVersion: GrammarVersion,
+  private val overwrite: Boolean,
+  private val names: Boolean,
+  private val silent: Boolean,
+  private val consoleWriter: Writer = System.out.writer(),
+  private val errWriter: Writer = System.err.writer(),
+) : CliCommand(CliBaseOptions()) {
+
+  private fun format(contents: String): String {
+    return Formatter().format(contents, grammarVersion)
+  }
+
+  private fun writeErr(error: String) {
+    if (silent) return
+    errWriter.write(error)
+    errWriter.appendLine()
+    errWriter.flush()
+  }
+
+  private fun write(message: String) {
+    if (silent) return
+    consoleWriter.write(message)
+    consoleWriter.appendLine()
+    consoleWriter.flush()
+  }
+
+  private fun allPaths(): Stream<Path> {
+    return paths
+      .distinct()
+      .stream()
+      .flatMap { path ->
+        if (path.isDirectory()) {
+          Files.walk(path).filter { it.extension == "pkl" || it.name == "PklProject" }
+        } else {
+          Stream.of(path)
+        }
+      }
+      .map { it.normalize().toAbsolutePath() }
+  }
+
+  override fun doRun() {
+    val status = Status(SUCCESS)
+
+    // Handle stdin input when path is "-"
+    if (paths.size == 1 && paths[0].toString() == "-") {
+      handleStdin(status)
+    } else {
+      handleFilePaths(status)
+    }
+
+    when (status.status) {
+      FORMATTING_VIOLATION -> {
+        if (!silent) {
+          writeErr("Formatting violations were found.")
+        }
+        // using CliTestException instead of CliException because we want full control on how to
+        // print errors
+        throw CliTestException("", status.status)
+      }
+      ERROR -> {
+        if (!silent) {
+          writeErr("An error occurred during formatting.")
+        }
+        throw CliTestException("", status.status)
+      }
+    }
+  }
+
+  private fun handleStdin(status: Status) {
     try {
-      return Formatter().format(contents, grammarVersion) to 0
+      val contents = System.`in`.bufferedReader().use { it.readText() }
+      val formatted = format(contents)
+
+      if (contents != formatted) {
+        status.update(FORMATTING_VIOLATION)
+      }
+
+      // For stdin, always output the formatted result (unless silent)
+      // The --names and -w flags don't make sense for stdin
```

Comment:
Makes sense, but, in this case, I'm thinking that we should throw if `-w` is found.